### PR TITLE
Nomi4

### PR DESCRIPTION
**Pull request by depappas-sifive**
## Purpose
Adding additional object model functionality

### Type of change
- New Feature

### CODE REVIEW

File: src/main/scala/diplomaticobjectmodel/DiplomaticObjectModelUtils.scala [hunk 1]
```diff
@@ -5,6 +5,12 @@ package freechips.rocketchip.diplomaticobjectmodel
 import java.io.{File, FileWriter}
 import java.lang.management.OperatingSystemMXBean
 
+import Chisel.{Data, SeqMem, Vec, log2Ceil}
+import chisel3.SyncReadMem
+import freechips.rocketchip.diplomacy.DTS.{Cells, fmtCell}
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/DiplomaticObjectModelUtils.scala]: Why are these being imported?

File: src/main/scala/diplomacy/Resources.scala [hunk 2]
```diff
@@ -361,10 +371,34 @@ trait BindingScope
...
+  }
+
+  /** Generate the ResourceBindingsMap which stores each device's ResourceBindings
+    *
+    * @return
+    */
+  def getResourceBindingsMap: ResourceBindingsMap = {
+    eval
+    ResourceBindingsMap(map = resourceBindings.reverse.groupBy(_._1.owner).mapValues(seq => ResourceBindings(
+        seq.groupBy(_._1.key).mapValues(_.map(z => Binding(z._2, z._3)).distinct))))
```
Comment (richardxia) [inline src/main/scala/diplomacy/Resources.scala]: This is the same expression that is in the `bindingTree` method. Can you refactor it such that `bindingTree` calls this?

File: src/main/scala/diplomacy/Resources.scala [hunk 3]
```diff
@@ -361,10 +371,34 @@ trait BindingScope
...
+    * @param rsb
+    */
+  def getResourceBindings(dev: Device, rsb: ResourceBindingsMap): Option[ResourceBindings] = {
+    if (rsb.map.contains(dev)) {
+      Some(rsb.map.lift(dev).getOrElse(ResourceBindings()))
+    }
+    else {
+      None
+    }
+  }
```
Comment (richardxia) [inline src/main/scala/diplomacy/Resources.scala]: Is this method being used anymore? I don't see it being called anywhere.

File: src/main/scala/diplomaticobjectmodel/DiplomaticObjectModelUtils.scala (hunk 1 above)
Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/DiplomaticObjectModelUtils.scala]: fixed

File: src/main/scala/diplomaticobjectmodel/DiplomaticObjectModelUtils.scala [hunk 4]
```diff
@@ -17,10 +23,110 @@ object DiplomaticObjectModelUtils {
...
+    Nil
+  }
+
+  def regFilter(name: String): Boolean = name == "reg" || name.take(4) == "reg/"
+
+  def regName(name: String): Option[String] = {
+    val keys = name.split("/")
+    require (keys.size >= 1 && keys.size <= 2 && keys(0) == "reg", s"Invalid reg name '${name}'")
+    if (keys.size == 1) None else Some(keys(1))
+  }
```
[the remark this answers is not in the document]
Comment (depappas-sifive) [reply src/main/scala/diplomaticobjectmodel/DiplomaticObjectModelUtils.scala]: Moved  regFilter() and regName() to an object.

File: src/main/scala/diplomacy/Resources.scala (hunk 2 above)
Comment (depappas-sifive) [reply src/main/scala/diplomacy/Resources.scala]: removed

File: src/main/scala/diplomacy/Resources.scala (hunk 3 above)
Comment (depappas-sifive) [reply src/main/scala/diplomacy/Resources.scala]: removed

**Pull request merged**